homeassistant/components/rabbitair/fan.py

```diff
@@ -0,0 +1,148 @@
+"""Support for Rabbit Air fan entity."""
+from __future__ import annotations
+
+from typing import Any
+
+from rabbitair import Mode, Model, Speed
+
+from homeassistant.components.fan import FanEntity, FanEntityFeature
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util.percentage import (
+    ordered_list_item_to_percentage,
+    percentage_to_ordered_list_item,
+)
+
+from .const import DOMAIN
+from .coordinator import RabbitAirDataUpdateCoordinator
+from .entity import RabbitAirBaseEntity
+
+SPEED_LIST = [
+    Speed.Silent,
+    Speed.Low,
+    Speed.Medium,
+    Speed.High,
+    Speed.Turbo,
+]
+
+PRESET_MODE_AUTO = "Auto"
+PRESET_MODE_MANUAL = "Manual"
+PRESET_MODE_POLLEN = "Pollen"
+
+PRESET_MODES = {
+    PRESET_MODE_AUTO: Mode.Auto,
+    PRESET_MODE_MANUAL: Mode.Manual,
+    PRESET_MODE_POLLEN: Mode.Pollen,
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up a config entry."""
+    coordinator: RabbitAirDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities([RabbitAirFanEntity(coordinator, entry)])
+
+
+class RabbitAirFanEntity(RabbitAirBaseEntity, FanEntity):
+    """Fan control functions of the Rabbit Air air purifier."""
+
+    _attr_supported_features = FanEntityFeature.PRESET_MODE | FanEntityFeature.SET_SPEED
+    _power: bool | None = None
+
+    def __init__(
+        self,
+        coordinator: RabbitAirDataUpdateCoordinator,
+        entry: ConfigEntry,
+    ) -> None:
+        """Initialize the entity."""
+        super().__init__(coordinator, entry)
+
+        if self._is_model(Model.MinusA2):
+            self._attr_preset_modes = list(PRESET_MODES)
+        elif self._is_model(Model.A3):
+            # A3 does not support Pollen mode
+            self._attr_preset_modes = [
+                k for k in PRESET_MODES if k != PRESET_MODE_POLLEN
+            ]
+
+        self._attr_speed_count = len(SPEED_LIST)
+
+        self._update_state()
```

Comment:
We should not be updating the state here. The coordinator should have the latest state and the entity would reflect that.
The coordinator will signal an update state on the entity when it refreshed/fetches data.

Reply:
Perhaps the name is poorly chosen and misleading. This is just an internal helper method that transforms the raw data received from the device API and fills in the corresponding class fields that store the current state of the device.